osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.UserInterface;
+using osu.Framework.Input.Events;
+using osuTK;
+
+namespace osu.Game.Overlays.BeatmapListing
+{
+    public abstract class BeatmapSearchMultipleSelectionFilterRow<T> : BeatmapSearchFilterRow<List<T>>
+    {
+        protected BeatmapSearchMultipleSelectionFilterRow(string headerName)
+            : base(headerName)
+        {
+        }
+
+        protected override Drawable CreateFilter() => CreateMultipleSelectionFilter();
+
+        protected abstract MultipleSelectionFilter CreateMultipleSelectionFilter();
+
+        protected abstract class MultipleSelectionFilter : FillFlowContainer<MultipleSelectionFilterTabItem>, IHasCurrentValue<List<T>>
+        {
+            private readonly BindableWithCurrent<List<T>> current = new BindableWithCurrent<List<T>>();
+
+            public Bindable<List<T>> Current
+            {
+                get => current.Current;
+                set => current.Current = value;
+            }
+
+            protected MultipleSelectionFilter()
+            {
+                Anchor = Anchor.BottomLeft;
+                Origin = Anchor.BottomLeft;
+                RelativeSizeAxes = Axes.X;
+                Height = 15;
+                Spacing = new Vector2(10, 0);
+                AddRange(CreateItems());
+
+                foreach (var item in Children)
+                    item.StateUpdated += updateBindable;
+            }
+
+            protected abstract MultipleSelectionFilterTabItem[] CreateItems();
```

Comment:
What is the point? All you care about is the underlying models, not the drawables themselves. How about a

```suggestion
            protected abstract T[] CreateItems();
```

or, even if you decided you want to customise the display text for each model, then

```suggestion
            protected abstract (T model, string description)[] CreateItems();
```